More build stages

UziTech opened the pull request:
**Marked version:** latest

## Description

This adds separate build stages for unit, spec, and integration tests as well as a minify build step which is only run when the master branch is pushed to.

The minify build step requires a github token be stored on travis (this should be done securely) and only given `public_repo` access

I'm not sure what we want to do with the different node versions. I don't think we need to run all tests against all node versions.

Later we can add a browser test stage.

![image](https://user-images.githubusercontent.com/97994/36982512-9e7fc332-2055-11e8-8c90-759a9d5056b4.png)


## Contributor

- [ ] Test(s) exist to ensure functionality and minimize regresstion (if no tests added, list tests covering this PR); or,
- [x] no tests required for this PR.
- [ ] If submitting new feature, it has been documented in the appropriate places.

## Committer

In most cases, this should be a different person than the contributor.

- [ ] Draft GitHub release notes have been updated.
- [ ] CI is green (no forced merge required).
- [ ] Merge PR

davisjam commented:
> I don't think we need to run all tests against all node versions.

I think we should run tests against any node versions we support. For example, if `const/let` are put into the source, that will break backwards compatibility and we would like to know.

UziTech commented:
I agree, but I don't think we need to run EVERY test against EVERY node version we support.

In this PR I have spec and unit tests only running against node lts and integration tests running against node v0.10, v4, lts, and latest

UziTech commented:
@styfle `marked.min.js` should never be manually edited so there should never be merge conflicts and `marked.min.js` should always be in sync with `marked.js`

UziTech commented:
also npm is not the only package manager that marked is in. So having `marked.min.js`  only as part of the npm publish would not be a good idea

styfle commented:
> should never be manually edited so there should never be merge conflicts

Generated code can still have merge conflicts.
This is why outputs/binaries are rarely committed to source control.

> npm is not the only package manager that marked is in

Do we have the other package managers listed anywhere? Who is publishing to them?

UziTech commented:
I agree the min file should just be in the release but since it has been part of the repo for a long time i'm guessing it will break something if we remove it now. Maybe in v2.0 we should remove it.

> Do we have the other package managers listed anywhere?

I'm not sure if they are listed anywhere but here are the ones I know of:
  - [npm](https://www.npmjs.com/package/marked)
  - [cdnjs](https://cdnjs.com/libraries/marked)
  - [bower](https://bower.io/search/?q=marked)
  - [jsdelivr](https://www.jsdelivr.com/package/npm/marked)
  - [unpkg](http://unpkg.org/pkg/marked/)

styfle commented:
I know that `npm`, `jsdelivr`, `unpkg` all read from npm so we don't have to worry, its publish once and they all get updated.

Is `bower` still around? I was thinking they were encouraging users to [migrate away from bower](https://bower.io/blog/2017/how-to-migrate-away-from-bower/).

I have no idea how to publish to `cdnjs`, it seems its completely manual.

styfle commented:
> Maybe in v2.0 we should remove it.

Thats fine

UziTech commented:
> Is `bower` still around?

They are recommending people move to yarn (which also uses npm registry) but I'm sure there will still be people using them for a while.

joshbruce commented:
@UziTech: If this doesn't break anything I'm fine merging it now. Just let me know.

UziTech commented:
I removed the "spec tests" and renamed "integration tests" to "spec tests"

In order for the minify step to work travis will need a `GITHUB_TOKEN` environment variable

![image](https://user-images.githubusercontent.com/97994/38582810-98bf87b4-3cd6-11e8-92ba-23f8a68e12cf.png)

The token only needs the "public_repo" permissions

![image](https://user-images.githubusercontent.com/97994/38582904-ed020fb8-3cd6-11e8-9daa-4127913bc6fb.png)

@joshbruce I'm not sure if you want to create a personal access token from your account otherwise I have seen organizations create a dummy account just for that.

UziTech commented:
Here is a test travis build showing the output for the minify job:
https://travis-ci.org/UziTech/marked/builds/349322194#L421

The GITHUB_TOKEN variable won't be leaked in the build output
![image](https://user-images.githubusercontent.com/97994/38583882-1869d408-3cda-11e8-8a2f-78bada1f02bb.png)

styfle commented:
I'm still going to voice my opinion that travis should not be committing code.
This will very likely cause merge conflicts for every PR because the minified file is a single line.

UziTech commented:
The minify job is only run when code is pushed to markedjs/master so marked.min.js isn't changed in PRs (and should never be manually changed in PRs)

UziTech commented:
If this does cause problems, it would be trivial to remove the job and go back to the way we are updating `marked.min.js` now. So including the job now can only improve things.

joshbruce commented:
@UziTech: Can you hook me up with instructions for the keys you need (it's not something I do on the regular pretty sure it's not the SSH ones that we need here). I'll decide what to do about accounts after that.

UziTech commented:
1. goto https://github.com/settings/tokens
2. click <kbd>Generate new token</kbd>
3. Enter a "Token description"
4. Check "public_repo"
5. Click <kbd>Generate token</kbd>
6. Copy the token
![image](https://user-images.githubusercontent.com/97994/38629869-dbeea976-3d7a-11e8-9115-57f11746a0de.png)
7. Goto https://travis-ci.org/markedjs/marked/settings
8. Enter the token under "Environment Variables"
![image](https://user-images.githubusercontent.com/97994/38629999-543d121e-3d7b-11e8-9366-a58f00bb0f68.png)
9. Click <kbd>Add</kbd>
10. Do what comes before Part-B 🎉

joshbruce commented:
@UziTech: Think I got it right. I created an account using the markedjs.org domain. It's just a bot with a forward to me at present; so, we'll figure stuff out as we progress.

UziTech commented:
Merge this and see what happens 😁

UziTech commented:
@joshbruce I think you have to add markedjsbot to the markedjs organization and git him push permissions

UziTech commented:
the error given is:

```
Permission to markedjs/marked.git denied to markedjsbot.
```

so it seems like the code worked, markedjsbot just doesn't have permission to push code to markedjs/marked

joshbruce commented:
Doh! And done.

UziTech commented:
Still getting the same error 😕 

Try making sure the permissions are correct then go to [travis](https://travis-ci.org/markedjs/marked) and click retry on the minify job

![image](https://user-images.githubusercontent.com/97994/38696931-24690fe4-3e56-11e8-8e6c-5344d6af12f7.png)

joshbruce commented:
Think `markdjsbot` had to be listed as an owner (not just a member of the org)??

styfle commented:
The minify step stopped working after I changed the settings for requiring approvers.

![image](https://user-images.githubusercontent.com/229881/39087729-6082c032-4573-11e8-977e-b6ce521bbc1e.png)


Which one do we think is more important:
Auto commit minify step or 2 required approvers?

joshbruce commented:
Is there no way to have both. My two cents is that we have ways around either one not being available; so, neither is a dyed in the wool must have. 

For releases having the minification happen on Travis cuts down work for me and minimizes the possibility of being out of sync.

UziTech commented:
I say the 2 approvers be just an unwritten rule

styfle commented:
I had to disable all the policies...leaving the `master` branch unprotected in order to allow travis to commit to master.
